CMake fails to find PETSc with pkg-config 0.29.2 and CPATH defined

<!--
Use this issue if you found a bug in the preCICE library. You can also open issues in [preCICE adapters and tutorials](https://github.com/precice).
For general user support, please use the [Discourse forum](https://precice.discourse.group/) or the [Gitter chatroom](https://gitter.im/precice/Lobby).
-->

**Describe your setup**

Initially discovered when installing preCICE using EasyBuild https://github.com/easybuilders/easybuild-easyconfigs/pull/12270
but reproduced in a Docker image of Ubuntu 20.10 Groovy

Operating system (e.g. Linux distribution and version): _Ubuntu 20.10_
preCICE Version: 2.2.0


**Describe the problem**

- The FindPETSc.cmake delivered with preCICE heavily relies on `pkg-config` to find PETSc.
- With new version 0.29.2, `pkg-config` will not return the required flags `-I` if already set in the `CPATH` environment variable (https://github.com/frida/pkg-config/commit/4ade8fc90115c4932bc198a500a1efd13f243cd4)
- So if `CPATH` contains the include path to the PETSc headers, preCICE will not be able to find PETSc and fails with error

```
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.29.2") 
-- Detecting additional PETSc prefixes
-- pkg-config will search the following paths:
-- Checking for module 'PETSc>=3.12'
--   Found PETSc, version 3.13.4
CMake Error at /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:146 (message):
  Could NOT find PETSc (missing: PETSc_INCLUDE_DIRS) (found suitable version
  "3.13.4", minimum required is "3.12")
Call Stack (most recent call first):
  /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:393 (_FPHSA_FAILURE_MESSAGE)
  cmake/modules/FindPETSc.cmake:229 (find_package_handle_standard_args)
  CMakeLists.txt:216 (find_package)
```


**Step To Reproduce**

1. Start a Docker of Ubuntu 20.10 because it provides `pkg-config` 0.29.2
```
docker run -ti ubuntu:groovy
```

2. Install requirements for preCICE
```
apt install wget cmake build-essential libboost-all-dev libeigen3-dev libxml2-dev python3-numpy petsc-dev
```

3. Download preCICE
```
wget https://github.com/precice/precice/archive/v2.2.0.tar.gz
tar -xvzf v2.2.0.tar.gz
mkdir build
```

4. Run CMake without `CPATH` -> It works
```
cd build
cmake ../precice-2.2.0/ -DPRECICE_PETScMapping=ON
```
```
# This commands returns the correct include path
pkg-config --cflags-only-I PETSc
```

5. Run CMake with `CPATH` defined -> It fails
```
# clean build directory first
export CPATH=/usr/lib/petscdir/petsc3.13/x86_64-linux-gnu-real/include
cmake ../precice-2.2.0/ -DPRECICE_PETScMapping=ON
```
```
# Reason: this command returns nothing
pkg-config --cflags-only-I PETSc
```



**Additional context**

This was initially discovered when using EasyBuild because the module created for PETSc sets the `CPATH` to the PETSc headers. A workaround has been implemented in https://github.com/easybuilders/easybuild-easyconfigs/pull/12270 to disable this new behavior of `pkg-config`.

I believe that this bug comes from the FindPETSc.cmake shipped with preCICE because libXML2 is impacted in the same way, but CMake still manage to detect it properly. 
A bug fix could probably take inspiration from the way it is done for libXML2 https://github.com/Kitware/CMake/blob/master/Modules/FindLibXml2.cmake

--- a/cmake/modules/FindPETSc.cmake
+++ b/cmake/modules/FindPETSc.cmake
@@ -143,9 +143,17 @@ if(PKG_CONFIG_FOUND)
     endif()
   endif()
 
+  # Set PKG_CONFIG_ALLOW_SYSTEM_CFLAGS
+  set(_petsc_prev_allow_system_cflags $ENV{PKG_CONFIG_ALLOW_SYSTEM_CFLAGS})
+  set(ENV{PKG_CONFIG_ALLOW_SYSTEM_CFLAGS} 1)
+
   # Use pkg-config to find PETSc
   set(PKG_CONFIG_USE_CMAKE_PREFIX_PATH "YES")
   pkg_check_modules(PC_PETSc ${_petsc_quiet_arg} "PETSc${_pkg_version_spec}")
+
+  # Restore/Reset PKG_CONFIG_USE_CMAKE_PREFIX_PATH
+  set(ENV{PKG_CONFIG_ALLOW_SYSTEM_CFLAGS} ${_petsc_prev_allow_system_cflags})
+
   unset(PKG_CONFIG_USE_CMAKE_PREFIX_PATH)
   unset(_pkg_version_spec)
 

--- a/docs/changelog/993.md
+++ b/docs/changelog/993.md
@@ -1,0 +1,1 @@
+- Fixed FindPETSc to also find PETSc if the include directory is in CPATH, which occured with some versions of `pkg-config`.

